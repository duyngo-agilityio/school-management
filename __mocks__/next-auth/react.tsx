import { createContext } from "react";

export const useSession = () => ({
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
});

export type Session = {
  data: {
    user: {
      token: string;
      email: string;
      id: string;
      name: string;
    };
    expires: string;
  };
  status: "loading" | "authenticated" | "unauthenticated";
  update: (data?: Session) => Promise<Session | null>;
};

const INITIAL_SESSION: Session = {
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
};

const SessionContext = createContext(INITIAL_SESSION);

export const SessionProvider = ({ children }: React.PropsWithChildren) => {
  <SessionContext.Provider value={INITIAL_SESSION}>
    {children}
  </SessionContext.Provider>;
};
