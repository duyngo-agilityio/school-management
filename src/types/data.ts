export interface IStudent {
  id: string;
  fullName: string;
  gender: number;
  age: number;
  className: string;
  password?: string;
  email?: string;
  description?: string;
  phoneNumber?: string;
  avatar?: string;
}

export interface ITeacher {
  id: string;
  fullName: string;
  gender: number;
  age: number;
  className: string;
  subject?: string;
  password?: string;
  email?: string;
  description?: string;
  phoneNumber?: string;
  avatar?: string;
}

export type TDataType = IStudent | ITeacher;

export type APIResponse<T> = {
  isSuccess: boolean;
  data: T | null;
};

export type TSearchParams = {
  q: string;
};