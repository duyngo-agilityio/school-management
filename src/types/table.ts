export interface IFiledNameColumns {
  field: string;
  headerName: string;
}
