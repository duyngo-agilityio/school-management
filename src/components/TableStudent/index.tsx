'use client';

// Libs
import Link from 'next/link';
import { useParams } from 'next/navigation';
import { useState } from 'react';

// Components
import { Box, Button, useDisclosure, useToast } from '@chakra-ui/react';
import Table from '../Table';
import Profile from '../Profile';
import StudentModal from '../Modal/Student/Modal';
import ConfirmModal from '../Modal/ConfirmModal';

// Constants
import {
  COLUMNS,
  customColors,
  ERROR_MESSAGES,
  ROUTES,
  SUCCESS_MESSAGES,
} from '@/constants';

// Types
import { IStudent } from '@/types';

// Icons
import PenIcon from '@/icons/PenIcon';
import TrashIcon from '@/icons/TrashIcon';

// Services
import { deleteStudent } from '@/actions';

interface TableStudentProps {
  data: IStudent[];
}

const TableStudent = ({ data }: TableStudentProps) => {
  const {
    isOpen: isOpenStudentModal,
    onOpen: onOpenStudentModal,
    onClose: onCloseStudentModal,
  } = useDisclosure();
  const {
    isOpen: isOpenConfirmModal,
    onOpen: onOpenConfirmModal,
    onClose: onCloseConfirmModal,
  } = useDisclosure();
  const toast = useToast();

  const param = useParams<{ id: string }>();

  const [idItem, setIdItem] = useState<string>('');

  const handleShowConfirmModal = (id: string) => () => {
    setIdItem(id);
    onOpenConfirmModal();
  };

  const handleSubmitDelete = async () => {
    const isSuccess = await deleteStudent(idItem);

    if (isSuccess) {
      onCloseConfirmModal();

      toast({
        title: SUCCESS_MESSAGES.DELETE_TEACHER,
        status: 'success',
      });
    } else {
      toast({
        title: ERROR_MESSAGES.DELETE_TEACHER,
        status: 'error',
      });
    }
  };

  const COLUMNS_STUDENT = [
    {
      field: COLUMNS.FIELDS.NAME,
      headerName: COLUMNS.HEADER_NAME.NAME,
      render: ({ fullName, avatar, id }: IStudent) => {
        return (
          <Link href={`${ROUTES.STUDENT}/${id}`} prefetch>
            <Profile src={avatar} title={fullName} variant="xs" />
          </Link>
        );
      },
    },
    {
      field: COLUMNS.FIELDS.ID,
      headerName: COLUMNS.HEADER_NAME.STUDENT_ID,
    },
    {
      field: COLUMNS.FIELDS.EMAIL,
      headerName: COLUMNS.HEADER_NAME.EMAIL,
    },
    {
      field: COLUMNS.FIELDS.CLASS,
      headerName: COLUMNS.HEADER_NAME.CLASS,
    },
    {
      field: COLUMNS.FIELDS.GENDER,
      headerName: COLUMNS.HEADER_NAME.GENDER,
      render: ({ gender }: IStudent) => {
        const renderGender = gender === 1 ? 'Male' : 'Female';

        return <Box>{renderGender}</Box>;
      },
    },
    {
      field: COLUMNS.FIELDS.ACTION,
      headerName: COLUMNS.HEADER_NAME.ACTION,
      render: ({ id }: IStudent) => {
        const active = param.id === id;

        return (
          <>
            <Button variant="none" mr="10px" onClick={onOpenStudentModal}>
              <PenIcon
                stroke={active ? customColors.pure : customColors.emperor}
              />
            </Button>
            <Button variant="none" onClick={handleShowConfirmModal(id)}>
              <TrashIcon
                stroke={active ? customColors.pure : customColors.emperor}
              />
            </Button>
          </>
        );
      },
    },
  ];

  return (
    <>
      {isOpenStudentModal && <StudentModal onClose={onCloseStudentModal} />}
      {isOpenConfirmModal && (
        <ConfirmModal
          isOpen={isOpenConfirmModal}
          onClose={onCloseConfirmModal}
          onSubmit={handleSubmitDelete}
          title="Delete Student"
          subTitle="Are you sure you want to delete this Student?"
        />
      )}
      <Table columns={COLUMNS_STUDENT} data={data} />
    </>
  );
};

export default TableStudent;
