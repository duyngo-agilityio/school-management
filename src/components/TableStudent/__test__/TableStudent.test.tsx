import { render } from '@testing-library/react';
import TableStudent from '@/components/TableStudent';
import { IStudent } from '@/types';
import { useParams } from 'next/navigation';

const mockData: IStudent[] = [
  {
    id: '1',
    fullName: 'John Doe',
    avatar: '/path/to/avatar1.jpg',
    email: 'johndoe@example.com',
    className: '10A',
    gender: 1,
    age: 23,
  },
  {
    id: '2',
    fullName: 'Jane Smith',
    avatar: '/path/to/avatar2.jpg',
    email: 'janesmith@example.com',
    className: '11B',
    gender: 0,
    age: 20,
  },
];

jest.mock('next/navigation', () => ({
  useParams: jest.fn(),
  useRouter: () => ({ push: jest.fn() }),
}));

describe('TableStudent Component', () => {
  beforeEach(() => {
    (useParams as jest.Mock).mockReturnValue({ id: '1' }); // Set a mock id
  });

  it('should render correctly and match snapshot', () => {
    const { container } = render(<TableStudent data={mockData} />);

    expect(container).toMatchSnapshot();
  });
});
