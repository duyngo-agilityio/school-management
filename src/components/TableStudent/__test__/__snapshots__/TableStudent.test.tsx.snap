// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableStudent Component should render correctly and match snapshot 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-0"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-ujzx5o"
          >
            Name
          </th>
          <th
            class="css-ujzx5o"
          >
            Student ID
          </th>
          <th
            class="css-ujzx5o"
          >
            Email Address
          </th>
          <th
            class="css-ujzx5o"
          >
            Class
          </th>
          <th
            class="css-ujzx5o"
          >
            Gender
          </th>
          <th
            class="css-ujzx5o"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-1t3b6fg"
        >
          <td
            class="css-4e3ndm"
          >
            <a
              href="/students/1"
            >
              <div
                class="css-0"
              >
                <span
                  class="chakra-avatar css-1ruhx4w"
                >
                  <svg
                    aria-label=" avatar"
                    class="chakra-avatar__svg css-16ite8i"
                    role="img"
                    viewBox="0 0 128 128"
                  >
                    <path
                      d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                      fill="currentColor"
                    />
                    <path
                      d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <p
                  class="chakra-text css-1isemmb"
                >
                  John Doe
                </p>
              </div>
            </a>
          </td>
          <td
            class="css-4e3ndm"
          >
            1
          </td>
          <td
            class="css-4e3ndm"
          >
            johndoe@example.com
          </td>
          <td
            class="css-4e3ndm"
          >
            10A
          </td>
          <td
            class="css-4e3ndm"
          >
            <div
              class="css-0"
            >
              Male
            </div>
          </td>
          <td
            class="css-4e3ndm"
          >
            <button
              class="chakra-button css-1uiavz5"
              type="button"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 25 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.5499 15.5999L9.7499 19.1999M12.7499 22.1999H22.3499M4.9499 15.5999L16.7813 3.35533C18.0552 2.08143 20.1206 2.08143 21.3945 3.35533C22.6684 4.62923 22.6684 6.69463 21.3945 7.96853L9.1499 19.7999L3.1499 21.5999L4.9499 15.5999Z"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
            <button
              class="chakra-button css-4xx2wk"
              type="button"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 6.17647H20M10 16.7647V10.4118M14 16.7647V10.4118M16 21H8C6.89543 21 6 20.0519 6 18.8824V7.23529C6 6.65052 6.44772 6.17647 7 6.17647H17C17.5523 6.17647 18 6.65052 18 7.23529V18.8824C18 20.0519 17.1046 21 16 21ZM10 6.17647H14C14.5523 6.17647 15 5.70242 15 5.11765V4.05882C15 3.47405 14.5523 3 14 3H10C9.44772 3 9 3.47405 9 4.05882V5.11765C9 5.70242 9.44772 6.17647 10 6.17647Z"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr
          class="css-bgbn71"
        >
          <td
            class="css-4e3ndm"
          >
            <a
              href="/students/2"
            >
              <div
                class="css-0"
              >
                <span
                  class="chakra-avatar css-1ruhx4w"
                >
                  <svg
                    aria-label=" avatar"
                    class="chakra-avatar__svg css-16ite8i"
                    role="img"
                    viewBox="0 0 128 128"
                  >
                    <path
                      d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                      fill="currentColor"
                    />
                    <path
                      d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <p
                  class="chakra-text css-1isemmb"
                >
                  Jane Smith
                </p>
              </div>
            </a>
          </td>
          <td
            class="css-4e3ndm"
          >
            2
          </td>
          <td
            class="css-4e3ndm"
          >
            janesmith@example.com
          </td>
          <td
            class="css-4e3ndm"
          >
            11B
          </td>
          <td
            class="css-4e3ndm"
          >
            <div
              class="css-0"
            >
              Female
            </div>
          </td>
          <td
            class="css-4e3ndm"
          >
            <button
              class="chakra-button css-1uiavz5"
              type="button"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 25 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.5499 15.5999L9.7499 19.1999M12.7499 22.1999H22.3499M4.9499 15.5999L16.7813 3.35533C18.0552 2.08143 20.1206 2.08143 21.3945 3.35533C22.6684 4.62923 22.6684 6.69463 21.3945 7.96853L9.1499 19.7999L3.1499 21.5999L4.9499 15.5999Z"
                  stroke="#4F4F4F"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
            <button
              class="chakra-button css-4xx2wk"
              type="button"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 6.17647H20M10 16.7647V10.4118M14 16.7647V10.4118M16 21H8C6.89543 21 6 20.0519 6 18.8824V7.23529C6 6.65052 6.44772 6.17647 7 6.17647H17C17.5523 6.17647 18 6.65052 18 7.23529V18.8824C18 20.0519 17.1046 21 16 21ZM10 6.17647H14C14.5523 6.17647 15 5.70242 15 5.11765V4.05882C15 3.47405 14.5523 3 14 3H10C9.44772 3 9 3.47405 9 4.05882V5.11765C9 5.70242 9.44772 6.17647 10 6.17647Z"
                  stroke="#4F4F4F"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
